### PR TITLE
pipeline module, improvements may be needed

### PR DESCRIPTION
**Pull request by Szymon-Gut**

### CODE REVIEW

**Comment by maciejors** [comment]
ratio_classes looks good to me now. good job on that

`train_loader` and `test_loader` were missing from `Pipeline.__slots__` which was causing an error. I added them there.

Now I'm getting a new error:
image
not sure what causes it, perhaps the parameter values of `n_queries=10, init_epochs=100, epochs_per_query=100` that I'm passing here don't make sense but idk.

edit: This code should reproduce the error (I run it in a notebook):
```py
%load_ext autoreload
%autoreload 2

import sys
sys.path.append("../..")

import torch
import torch.nn.functional as F
from torch.utils.data import DataLoader

import src.modules.data_module as dm
import src.modules.learner_module as lm
import src.models as models
import src.modules.pipeline as pp

device = 'cpu'

# unified test set
test_size = 1000
ds_test = dm.ActiveDataset('mnist', 0, test_size, balanced_split=False)
test_loader = DataLoader(ds_test.test_set, batch_size=128)

init_samples = 100
train_size = 10000
loss_fn = F.cross_entropy

# init learner
model = models.MNISTClassifier().to(device)
learner = lm.ActiveLearner(model, device)
optimizer = torch.optim.Adam(params=model.parameters())

# init set
ds = dm.ActiveDataset('mnist', train_size, 0, ratio_labeled=init_samples/train_size, balanced_split=False)

pipe = pp.Pipeline()
pipe.fit_transform(data_set=ds, learner=learner, optimizer=optimizer, loss_function=loss_fn, n_queries=5, init_epochs=1000, epochs_per_query=1000, query_batch_size=500)
```

**Comment by krezelj** [comment]
I believe it might be the case that you have test_size set to 0 and use the generated (empty) test set to create a test_loader that is then passed to `learner.fit()` method. No validation loss is calculated and so `i` is not declared in the loop which causes the error. This concerns the learner module, I'll create an issue for it.

EDIT: Issue #16

File: src/modules/pipeline.py
Comment (krezelj) [inline src/modules/pipeline.py:1]: As per my comment on the group chat the Pipeline should be simpler:
- use `__init__` to set variables like `ds : ActiveDataset`, `learner: ActiveLearner`, `n_queries`, `queries_per_batch` etc.
    - might be useful to implement `PipelineSettings` class that holds all settings (apart from `ds` and `learner` and allows for writing/reading to/from .json file 
- implement `run` method that runs the pipeline on the set variables and outputs `em.Session` object.
    - apart from `em.Session` it might also return a dict of stats containing things like `t_loss_history`, `v_loss_history`, `accuracy_history` etc.
- get rid of `fit`, `transform`, `fit_transform` methods

File: src/modules/data_module.py
Comment (krezelj) [inline src/modules/data_module.py:1]: Looks good

**Review by krezelj — CHANGES_REQUESTED** [review]

**Comment by maciejors** [comment]
`Pipeline` class is now implemented with all the changes requested. The `run()` method mostly reuses the code that was earlier inside the old `transform()` method.

I also added `PipelineSettings` class as suggested. It supports exporting/importing to/from dict and json files.

File: src/modules/pipeline.py [hunk 1]
```diff
@@ -1,10 +1,183 @@
...
+        session = em.Session(dataset=self.dataset, learner=self.learner)
+
+        for i in range(self.settings.n_queries):
+            print(f"Iteration: {i+1}")
+
+            # generate queries
+            unlabeled_loader = DataLoader(self.dataset.unlabeled_set, batch_size=128, shuffle=False)
+            queries, uncertainty = self.learner.generate_query(
+                unlabeled_loader,
+                criterion='entropy',
```
Comment (krezelj) [inline src/modules/pipeline.py]: Make this a parameter of PipelineSettings. Possible values: `confidence`, `margin`, `entropy`, `random`.

File: src/modules/pipeline.py [hunk 2]
```diff
@@ -1,10 +1,183 @@
...
+                batch_size=self.settings.query_batch_size,
+            )
+            # label queries
+            self.dataset.get_label_by_idx(queries, move_sample=True)
+            print(f"Max uncertainty: {uncertainty}")
```
Comment (krezelj) [inline src/modules/pipeline.py]: Add `verbose` as parameter of run default is 0. If `verbose>1` display max uncertainites.

File: src/modules/pipeline.py [hunk 3]
```diff
@@ -1,10 +1,183 @@
...
+            epochs=self.settings.init_epochs,
+            early_stopping=True,
+        )
+        accuracy_history = [self._calculate_accuracy(self.test_loader).item()]
+
+        # initialise the session
+        session = em.Session(dataset=self.dataset, learner=self.learner)
+
+        for i in range(self.settings.n_queries):
+            print(f"Iteration: {i+1}")
```
Comment (krezelj) [inline src/modules/pipeline.py]: Add `verbose` as parameter of run default is 0. If `verbose>0` display current iteration

File: src/modules/pipeline.py [hunk 4]
```diff
@@ -1,10 +1,183 @@
...
+            t_loss_history.extend(t_hist)
+            v_loss_history.extend(v_hist)
+            accuracy_history.append(self._calculate_accuracy(self.test_loader).item())
+
+        # pack loss/accuracy history arrays into one dict
+        stats_to_return = {
+            't_loss_history': t_loss_history,
+            'v_loss_history': v_loss_history,
+            'accuracy_history': accuracy_history,
+        }
```
Comment (krezelj) [inline src/modules/pipeline.py]: Add `f1_score_history` as one of the returned stats. Add parameters to `run` function:
- `calculate_accuracy=False`
    - only calculate accuracy if True (to speed up computation)
- `calculate_f1_score=False`
    - same as above

File: src/modules/pipeline.py [hunk 5]
```diff
@@ -1,10 +1,183 @@
...
+            outputs,
+            test_loader.dataset.dataset.targets[test_loader.dataset.indices]
+        )
+
+    def _calculate_accuracy(self, test_loader):
+        outputs = self.learner.predict(test_loader)
+        return multiclass_accuracy(
+            outputs,
+            test_loader.dataset.dataset.targets[test_loader.dataset.indices]
+        )
```
Comment (krezelj) [inline src/modules/pipeline.py]: Pipeline already has slot `test_loader` that is set in `__init__`. When calling this function we pass `self.test_loader` anyway so I would get rid of this parameter and use `self.test_loader` inside the functions instead.

```suggestion
    def _calculate_f1_score(self):
        outputs = self.learner.predict(self.test_loader)
        return multiclass_f1_score(
            outputs,
            self.test_loader.dataset.dataset.targets[self.test_loader.dataset.indices]
        )

    def _calculate_accuracy(self):
        outputs = self.learner.predict(self.test_loader)
        return multiclass_accuracy(
            outputs,
            self.test_loader.dataset.dataset.targets[self.test_loader.dataset.indices]
        )
```

**Review by krezelj — CHANGES_REQUESTED** [review]
Looks really good. Good job on the PipelineSettings. A few QOL suggestions.

**Comment by maciejors** [comment]
all done and tested

**Review by krezelj — APPROVED** [review]
Great thanks, approved
